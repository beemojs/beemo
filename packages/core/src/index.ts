/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Blueprint, PackageStructure, Path, PortablePath, Predicates } from '@boost/common';

export * from './constants';
export * from './contexts/ConfigContext';
export * from './contexts/Context';
export * from './contexts/DriverContext';
export * from './contexts/ScaffoldContext';
export * from './contexts/ScriptContext';
export * from './Driver';
export * from './Script';
export * from './Tool';
export * from './types';

export { Path };
export type { Blueprint, PackageStructure, PortablePath, Predicates };
