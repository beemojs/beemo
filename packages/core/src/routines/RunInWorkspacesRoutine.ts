import Graph from '@beemo/dependency-graph';
import { Blueprint, PackageStructure,Predicates, WorkspacePackage } from '@boost/common';
import { PooledPipeline,Routine } from '@boost/pipeline';
import { stripAnsi, style } from '@boost/terminal';
import Context from '../contexts/Context';
import isPatternMatch from '../helpers/isPatternMatch';
import { ExecutionError, RoutineOptions } from '../types';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyRoutine = Routine<any, any>;

export interface RunInWorkspacesContextArgs {
  concurrency: number;
  graph: boolean;
  workspaces: string;
}

const MAX_ERROR_LINES = 5;

export default abstract class RunInWorkspacesRoutine<
  Ctx extends Context<RunInWorkspacesContextArgs>
> extends Routine<unknown, unknown, RoutineOptions> {
  routines: AnyRoutine[] = [];

  workspacePackages: WorkspacePackage[] = [];

  blueprint({ instance }: Predicates): Blueprint<RoutineOptions> {
    return {
      // @ts-expect-error We cant import Tool because of cycles
      tool: instance().required().notNullable(),
    };
  }

  async execute(context: Ctx): Promise<unknown> {
    const { args, workspaces } = context;
    const { tool } = this.options;

    // Determine packages to run plugins in
    if (args.options.workspaces) {
      if (!workspaces || workspaces.length === 0) {
        throw new Error(tool.msg('errors:workspacesNotEnabled', { arg: args.options.workspaces }));
      }

      this.workspacePackages = tool.project.getWorkspacePackages();

      this.getFilteredWorkspacePackages(context).forEach((pkg) => {
        this.pipeRoutine(context, pkg.metadata.packageName, pkg.metadata.packagePath);
      });
    } else {
      this.pipeRoutine(context);
    }

    const value = await this.getInitialValue(context);
    const batches = this.orderByWorkspacePriorityGraph(context);
    const allErrors: Error[] = [];
    const allResults: unknown[] = [];
    const concurrency =
      context.args.options.concurrency || tool.config.execute.concurrency || undefined;

    // eslint-disable-next-line no-restricted-syntax
    for (const batch of batches) {
      const pipeline = batch.reduce(
        (pl, routine) => pl.add(routine),
        new PooledPipeline(context, value, { concurrency }),
      );

      // eslint-disable-next-line no-await-in-loop
      const { errors, results } = await pipeline.run();

      allResults.push(...results);

      if (errors.length > 0) {
        allErrors.push(...errors);
      }
    }

    if (allErrors.length > 0) {
      this.formatAndThrowErrors(allErrors);
    }

    // Not running in workspaces, so return value directly
    return context.args.options.workspaces ? allResults : allResults[0];
  }

  /**
   * When a list of errors are available, concatenate them and throw a new error.
   */
  formatAndThrowErrors(errors: Error[]) {
    let message = this.options.tool.msg('errors:executeFailed');

    (errors as ExecutionError[]).forEach((error) => {
      let content = stripAnsi(error.stderr || error.stdout || '')
        .trim()
        .split('\n');

      // istanbul ignore next
      if (content.length >= MAX_ERROR_LINES) {
        const count = content.length - MAX_ERROR_LINES;

        content = content.slice(0, MAX_ERROR_LINES);

        if (count > 0) {
          content.push(this.options.tool.msg('errors:executeFailedMoreLines', { count }));
        }
      }

      message += '\n\n';
      message += style.reset.yellow(error.message);

      if (content.length > 0) {
        message += '\n';
        message += style.reset.gray(content.join('\n'));
      }
    });

    message += '\n';

    const error = new Error(message);

    // Inherit stack for easier debugging.
    if (errors.length === 1) {
      error.stack = String(errors[0].stack).split('\n').slice(1).join('\n');
    }

    throw error;
  }

  /**
   * Return a list of workspaces optionally filtered.
   */
  getFilteredWorkspacePackages(context: Ctx): WorkspacePackage[] {
    return this.workspacePackages.filter((pkg) =>
      isPatternMatch(pkg.package.name, context.args.options.workspaces),
    );
  }

  /**
   * Return the initial value for the pipeline.
   */
  getInitialValue(context: Ctx): unknown {
    return null;
  }

  /**
   * Group routines in order of which they are dependend on.
   */
  orderByWorkspacePriorityGraph(context: Ctx): AnyRoutine[][] {
    const enabled = context.args.options.graph || this.options.tool.config.execute.graph || false;

    if (!enabled || !context.args.options.workspaces) {
      return [this.routines];
    }

    // Create lookups
    const packages: PackageStructure[] = []; // Without metadata
    const metadata: { [name: string]: WorkspacePackage['metadata'] } = {}; // By package name

    this.workspacePackages.forEach((wsp) => {
      packages.push(wsp.package);
      metadata[wsp.package.name] = wsp.metadata;
    });

    // Batch based on packages
    const batchList = new Graph(packages).resolveBatchList();
    const batches: AnyRoutine[][] = [];

    batchList.forEach((batch) => {
      const routines = batch
        .map((pkg) => this.routines.find((route) => route.key === metadata[pkg.name].packageName))
        .filter(Boolean) as AnyRoutine[];

      if (routines.length > 0) {
        batches.push(routines);
      }
    });

    return batches;
  }

  /**
   * Pipe a routine for the entire project or a workspace package at the defined path.
   */
  abstract pipeRoutine(context: Ctx, packageName?: string, packageRoot?: string): void;
}
