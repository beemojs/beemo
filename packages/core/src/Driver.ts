import execa from 'execa';
import mergeWith from 'lodash/mergeWith';
import { Blueprint, isObject, optimal, Path, Predicates, predicates } from '@boost/common';
import { ConcurrentEvent, Event } from '@boost/event';
import { Plugin } from '@boost/plugin';
import {
  STRATEGY_COPY,
  STRATEGY_CREATE,
  STRATEGY_NATIVE,
  STRATEGY_NONE,
  STRATEGY_REFERENCE,
} from './constants';
import ConfigContext from './contexts/ConfigContext';
import DriverContext from './contexts/DriverContext';
import isClassInstance from './helpers/isClassInstance';
import Tool from './Tool';
import {
  Argv,
  BeemoTool,
  Driverable,
  DriverCommandOptions,
  DriverMetadata,
  DriverOptions,
  DriverOutput,
  DriverStrategy,
  Execution,
} from './types';

export default abstract class Driver<
    Config extends object = {},
    Options extends DriverOptions = DriverOptions
  >
  extends Plugin<BeemoTool, Options>
  implements Driverable {
  command: DriverCommandOptions = {};

  // Set after instantiation
  config!: Config;

  // Set after instantiation
  metadata!: DriverMetadata;

  // Set within a life-cycle
  tool!: Tool;

  output: DriverOutput = {
    stderr: '',
    stdout: '',
  };

  readonly onLoadProviderConfig = new Event<[ConfigContext, Path, Config]>('load-provider-config');

  readonly onLoadConsumerConfig = new Event<[ConfigContext, Config]>('load-consumer-config');

  readonly onMergeConfig = new Event<[ConfigContext, Config]>('merge-config');

  readonly onCreateConfigFile = new Event<[ConfigContext, Path, Config]>('create-config-file');

  readonly onCopyConfigFile = new Event<[ConfigContext, Path, Config]>('copy-config-file');

  readonly onReferenceConfigFile = new Event<[ConfigContext, Path, Config]>(
    'reference-config-file',
  );

  readonly onDeleteConfigFile = new Event<[ConfigContext, Path]>('delete-config-file');

  readonly onBeforeExecute = new ConcurrentEvent<[DriverContext, Argv]>('before-execute');

  readonly onAfterExecute = new ConcurrentEvent<[DriverContext, unknown]>('after-execute');

  readonly onFailedExecute = new ConcurrentEvent<[DriverContext, Error]>('failed-execute');

  static validate(driver: Driver) {
    const name = (isClassInstance(driver) && driver.constructor.name) || 'Driver';

    if (!isObject(driver.options)) {
      throw new Error(`\`${name}\` requires an options object.`);
    }
  }

  blueprint({ array, object, string }: Predicates): Blueprint<DriverOptions> {
    return {
      args: array(string()),
      dependencies: array(string()),
      env: object(string()),
      strategy: string(STRATEGY_NATIVE).oneOf<DriverStrategy>([
        STRATEGY_NATIVE,
        STRATEGY_CREATE,
        STRATEGY_REFERENCE,
        STRATEGY_COPY,
        STRATEGY_NONE,
      ]),
    };
  }

  bootstrap() {}

  startup(tool: Tool) {
    this.tool = tool;
    this.bootstrap();
  }

  /**
   * Special case for merging arrays.
   */
  doMerge(prevValue: unknown, nextValue: unknown): unknown {
    if (Array.isArray(prevValue) && Array.isArray(nextValue)) {
      return Array.from(new Set(prevValue.concat(nextValue)));
    }

    return undefined;
  }

  /**
   * Extract the error message when the driver fails to execute.
   */
  extractErrorMessage(error: { message: string }): string {
    return error.message.split('\n', 1)[0] || '';
  }

  /**
   * Format the configuration file before it's written.
   */
  formatConfig(data: Config): string {
    const content = JSON.stringify(data, null, 2);

    if (this.metadata.configName.endsWith('.js')) {
      return `module.exports = ${content};`;
    }

    return content;
  }

  /**
   * Return the module name without the Beemo namespace.
   */
  getName(): string {
    return this.name.split('-').pop()!;
  }

  /**
   * Return a list of user defined arguments.
   */
  getArgs(): Argv {
    return Array.isArray(this.options.args) ? this.options.args : [];
  }

  /**
   * Return a list of dependent drivers.
   */
  getDependencies(): string[] {
    const dependencies = Array.isArray(this.options.dependencies) ? this.options.dependencies : [];

    return [
      // Always required; configured by the driver
      ...this.metadata.dependencies,
      // Custom; configured by the consumer
      ...dependencies,
    ];
  }

  /**
   * Return a list of supported CLI options.
   */
  getSupportedOptions(): string[] {
    return [];
  }

  /**
   * Extract the current version of the installed driver via its binary.
   */
  getVersion(): string {
    const { bin, versionOption } = this.metadata;
    const version = (execa.sync(bin, [versionOption])?.stdout || '').trim();
    const match = version.match(/(\d+)\.(\d+)\.(\d+)/u);

    return match ? match[0] : '0.0.0';
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: Config, next: Config): Config {
    return mergeWith(prev, next, this.doMerge);
  }

  /**
   * Handle command failures according to this driver.
   */
  processFailure(error: Execution) {
    const { stderr = '', stdout = '' } = error;
    const out = (stderr || stdout).trim();

    if (out) {
      this.setOutput('stderr', out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  processSuccess(response: Execution) {
    const { stderr = '', stdout = '' } = response;

    this.setOutput('stderr', stderr.trim());
    this.setOutput('stdout', stdout.trim());
  }

  /**
   * Setup additional command options.
   */
  setCommandOptions(options: DriverCommandOptions): this {
    const blueprint: Blueprint<DriverCommandOptions> = {};
    const { shape, string } = predicates;

    Object.keys(options).forEach((key) => {
      blueprint[key] = shape({
        description: string().notEmpty().required(),
        type: string().oneOf<'boolean' | 'number' | 'string'>(['string', 'number', 'boolean']),
      });
    });

    this.command = optimal(options, blueprint, {
      name: this.constructor.name,
      unknown: true,
    });

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: Partial<DriverMetadata>): this {
    const { array, bool, string } = predicates;

    this.metadata = optimal(
      metadata,
      {
        bin: string()
          .match(/^[a-z]{1}[a-zA-Z0-9-]+$/u)
          .required(),
        configName: string().required(),
        configOption: string('--config'),
        configStrategy: string(STRATEGY_CREATE).oneOf([
          STRATEGY_CREATE,
          STRATEGY_REFERENCE,
          STRATEGY_COPY,
        ]),
        dependencies: array(string()),
        description: string(),
        filterOptions: bool(true),
        helpOption: string('--help'),
        title: string().required(),
        useConfigOption: bool(),
        versionOption: string('--version'),
        watchOptions: array(string()),
        workspaceStrategy: string(STRATEGY_REFERENCE).oneOf([STRATEGY_REFERENCE, STRATEGY_COPY]),
      },
      {
        name: this.constructor.name,
      },
    );

    return this;
  }

  /**
   * Store the raw output of the driver's execution.
   */
  setOutput(type: keyof DriverOutput, value: string): this {
    this.output[type] = value.trim();

    return this;
  }
}
