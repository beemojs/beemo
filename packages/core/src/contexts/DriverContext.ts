import ConfigContext from './ConfigContext';
import Driver from '../Driver';
import { Arguments, Argv, StdioType } from '../types';

export interface DriverContextOptions {
  concurrency: number;
  graph: boolean;
  stdio: StdioType;
  workspaces: string;
}

export default class DriverContext extends ConfigContext<DriverContextOptions> {
  // Name defined on the plugin (kebab case)
  driverName: string = '';

  // Nested list of argv for each parallel execution
  parallelArgv: Argv[] = [];

  // The primary driver that initiated the pipeline
  primaryDriver: Driver;

  constructor(args: Arguments<DriverContextOptions>, driver: Driver, parallelArgv: Argv[] = []) {
    super(args);

    this.driverName = driver.name;
    this.parallelArgv = parallelArgv;
    this.primaryDriver = driver;

    // Add primary driver to driver list
    this.drivers.add(driver);
  }

  /**
   * Add a parallel command with additional argv.
   */
  addParallelCommand(argv: Argv): this {
    this.parallelArgv.push(argv);

    return this;
  }
}
