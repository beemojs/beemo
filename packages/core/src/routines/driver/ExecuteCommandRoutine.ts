/* eslint-disable @typescript-eslint/member-ordering */

import execa, { ExecaError } from 'execa';
import glob from 'fast-glob';
import fs from 'fs-extra';
import isGlob from 'is-glob';
import merge from 'lodash/merge';
import { parse } from '@boost/args';
import { Bind, Blueprint, ExitError, Path, Predicates } from '@boost/common';
import { color } from '@boost/internal';
import { AnyWorkUnit, Routine, WaterfallPipeline } from '@boost/pipeline';
import { STRATEGY_COPY } from '../../constants';
import { DriverContext } from '../../contexts/DriverContext';
import { filterArgs, OptionMap } from '../../helpers/filterArgs';
import { formatExecReturn } from '../../helpers/formatExecReturn';
import { BatchStream } from '../../streams/BatchStream';
import type { Tool } from '../../Tool';
import { Argv, Execution, RoutineOptions } from '../../types';

const OPTION_PATTERN = /-?-[a-z0-9-]+(,|\s)/giu;

export interface ExecuteCommandOptions extends RoutineOptions {
	additionalArgv?: Argv;
	argv?: Argv;
	forceConfigOption?: boolean;
	packageRoot?: string;
}

export class ExecuteCommandRoutine extends Routine<unknown, unknown, ExecuteCommandOptions> {
	blueprint({ array, bool, instance, string }: Predicates): Blueprint<ExecuteCommandOptions> {
		return {
			additionalArgv: array(string()),
			argv: array(string()),
			forceConfigOption: bool(),
			packageRoot: string(),
			tool: instance<Tool>().required().notNullable(),
		};
	}

	async execute(context: DriverContext) {
		const { tool } = this.options;
		const { forceConfigOption, packageRoot } = this.options;
		const { metadata, options } = context.primaryDriver;

		let pipeline = new WaterfallPipeline(context).pipe(
			tool.msg('app:driverExecuteGatherArgs'),
			this.gatherArgs,
		);

		if (options.expandGlobs) {
			pipeline = pipeline.pipe(tool.msg('app:driverExecuteExpandGlob'), this.expandGlobPatterns);
		}

		if (metadata.filterOptions) {
			pipeline = pipeline.pipe(
				tool.msg('app:driverExecuteFilterOptions'),
				this.filterUnknownOptions,
			);
		}

		if (packageRoot && metadata.workspaceStrategy === STRATEGY_COPY) {
			pipeline = pipeline.pipe(
				tool.msg('app:driverExecuteCopyWorkspaceConfig'),
				this.copyConfigToWorkspacePackage,
			);
		} else if (metadata.useConfigOption || forceConfigOption) {
			pipeline = pipeline.pipe(
				tool.msg('app:driverExecuteIncludeConfigOption'),
				this.includeConfigOption,
			);
		}

		return pipeline.pipe(tool.msg('app:driverExecute'), this.runCommandWithArgs).run();
	}

	/**
	 * Capture output when output strategy is "stream" or "pipe".
	 */
	@Bind()
	captureOutput(context: DriverContext, stream: execa.ExecaChildProcess) {
		const { tool } = this.options;
		const { args, primaryDriver } = context;
		const { watchOptions } = primaryDriver.metadata;
		const isWatching = watchOptions.some((option) => {
			// Option
			if (option.startsWith('-')) {
				return !!context.getRiskyOption(option.replace(/^-{1,2}/u, ''));
			}

			// Param
			return args.params.includes(option);
		});

		const outHandler = (chunk: Buffer) => {
			tool.outStream?.write(String(chunk));
		};

		const errHandler = (chunk: Buffer) => {
			tool.errStream?.write(String(chunk));
		};

		if (isWatching) {
			const wait = 1000;

			stream.stdout!.pipe(new BatchStream({ wait })).on('data', outHandler);
			stream.stderr!.pipe(new BatchStream({ wait })).on('data', errHandler);

			return 'watch';
		}

		const strategy = primaryDriver.getOutputStrategy();

		if (strategy === 'stream' || strategy === 'pipe') {
			stream.stdout!.on('data', outHandler);
			stream.stderr!.on('data', errHandler);
		}

		return strategy;
	}

	/**
	 * When workspaces are enabled, some drivers require the config to be within each workspace,
	 * instead of being referenced from the root, so we need to copy it.
	 */
	@Bind()
	copyConfigToWorkspacePackage(context: DriverContext, argv: Argv): Argv {
		const { packageRoot } = this.options;

		this.debug('Copying config files to workspace');

		context.configPaths.forEach((config) => {
			fs.copyFileSync(config.path.path(), new Path(packageRoot, config.path.name()).path());
		});

		return argv;
	}

	/**
	 * Expand arguments that look like globs.
	 */
	@Bind()
	expandGlobPatterns(context: DriverContext, argv: Argv): Argv {
		const nextArgv: Argv = [];

		this.debug('Expanding glob patterns');

		argv.forEach((arg) => {
			if (!arg.startsWith('-') && isGlob(arg)) {
				const paths = glob
					.sync(arg, {
						cwd: String(context.cwd),
						onlyDirectories: false,
						onlyFiles: false,
					})
					.map((path) => new Path(path).path());

				this.debug(
					'  %s %s %s',
					arg,
					color.mute('->'),
					paths.length > 0 ? paths.join(', ') : color.mute(this.options.tool.msg('app:noMatch')),
				);

				nextArgv.push(...paths);
			} else {
				nextArgv.push(arg);
			}
		});

		return nextArgv;
	}

	/**
	 * Extract native supported options and flags from driver help output.
	 */
	async extractNativeOptions(context: DriverContext): Promise<OptionMap> {
		const driver = context.primaryDriver;
		const { env } = driver.options;
		const options = driver.getSupportedOptions();

		if (options.length > 0) {
			this.debug('Using supported options from driver');

			const nativeOptions: OptionMap = {};

			options.forEach((option) => {
				nativeOptions[option] = true;
			});

			return Promise.resolve(nativeOptions);
		}

		this.debug('Extracting native options from help output');

		const { stdout } = await this.executeCommand(
			driver.metadata.bin,
			driver.metadata.helpOption.split(' '),
			{
				env,
				preferLocal: true,
			},
		);

		const nativeOptions: OptionMap = {};
		const matches = stdout.match(OPTION_PATTERN) ?? [];

		matches.forEach((option) => {
			// Trim trailing comma or space
			nativeOptions[option.slice(0, -1)] = true;
		});

		return nativeOptions;
	}

	/**
	 * Filter unknown and or unsupported CLI options from the arguments passed to the CLI.
	 * Utilize the driver's help option/command to determine accurate options.
	 */
	@Bind()
	async filterUnknownOptions(context: DriverContext, argv: Argv): Promise<Argv> {
		this.debug('Filtering unknown command line options');

		const nativeOptions = await this.extractNativeOptions(context);
		const { filteredArgv, unknownArgv } = filterArgs(argv, {
			allow: nativeOptions,
		});

		if (unknownArgv.length > 0) {
			this.debug('Filtered args: %s', color.mute(unknownArgv.join(', ')));
		}

		return filteredArgv;
	}

	/**
	 * Gather arguments from all sources to pass to the driver.
	 */
	@Bind()
	gatherArgs(context: DriverContext): Argv {
		this.debug('Gathering arguments to pass to driver');

		const argv = [
			// Passed by the driver
			...this.getDriverArgs(context),
			// Passed on the command line
			...this.getCommandLineArgs(),
			// Passed with parallel "//" operator
			...this.getAdditionalArgs(),
		];

		// Since we combine multiple args, we need to rebuild this.
		// And we need to set this before we filter them.
		// And we need to be sure not to remove existing args.
		context.args = merge(
			{},
			parse(argv, {
				loose: true,
				options: {},
			}),
			context.args,
		);

		return argv;
	}

	/**
	 * Run some validation on additional/parallel args.
	 */
	getAdditionalArgs(): Argv {
		const argv = this.options.additionalArgv;

		this.debug.invariant(argv.length > 0, 'From parallel operator', argv.join(' '), 'No arguments');

		return argv;
	}

	/**
	 * Return args from the command line.
	 */
	getCommandLineArgs(): Argv {
		const { argv } = this.options;

		this.debug.invariant(argv.length > 0, 'From the command line', argv.join(' '), 'No arguments');

		return argv;
	}

	/**
	 * Return args from the primary driver.
	 */
	getDriverArgs(context: DriverContext): Argv {
		const argv = context.primaryDriver.getArgs();

		this.debug.invariant(
			argv.length > 0,
			'From driver "args" option',
			argv.join(' '),
			'No arguments',
		);

		return argv;
	}

	/**
	 * Include --config option if driver requires it (instead of auto-lookup resolution).
	 */
	@Bind()
	includeConfigOption(context: DriverContext, prevArgv: Argv): Argv {
		const { primaryDriver } = context;
		const configPath = context.findConfigByName(primaryDriver.metadata.configName);
		const argv = [...prevArgv];

		if (configPath && primaryDriver.metadata.configOption) {
			argv.push(primaryDriver.metadata.configOption, configPath.path.path());
		}

		this.debug('Including config option to args');

		return argv;
	}

	/**
	 * Execute the driver's command with the filtered arguments and handle the
	 * success and failures with the driver itself.
	 */
	@Bind()
	async runCommandWithArgs(
		context: DriverContext,
		argv: Argv,
		workUnit?: AnyWorkUnit,
	): Promise<Execution> {
		const driver = context.primaryDriver;
		const cwd = String(this.options.packageRoot || context.cwd);

		this.debug(
			'Executing command "%s %s" in %s',
			color.symbol(driver.metadata.bin),
			argv.join(' '),
			color.filePath(cwd),
		);

		await driver.onBeforeExecute.emit([context, argv]);

		try {
			const result = await this.executeCommand(driver.metadata.bin, argv, {
				cwd,
				env: driver.options.env,
				preferLocal: true,
				workUnit,
				wrap: (stream) => this.captureOutput(context, stream),
			});

			this.debug('  Success: %o', formatExecReturn(result));

			driver.processSuccess(result);

			await driver.onAfterExecute.emit([context, result]);

			return result;
		} catch (error: unknown) {
			const result = error as ExecaError;

			this.debug('  Failure: %o', formatExecReturn(result));
			this.debug('  Error message: %s', color.fail(result.message));

			if (result.name !== 'MaxBufferError') {
				driver.processFailure(result);
			}

			await driver.onFailedExecute.emit([context, result]);

			// https://nodejs.org/api/child_process.html#child_process_event_exit
			throw result.exitCode === null && result.signal === 'SIGKILL'
				? new ExitError('Out of memory!', 1)
				: new ExitError((driver.extractErrorMessage(result) || '').trim(), result.exitCode);
		}
	}
}
