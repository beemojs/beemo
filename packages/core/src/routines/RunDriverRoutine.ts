import DriverContext from '../contexts/DriverContext';
import ExecuteCommandRoutine, { ExecuteCommandOptions } from './driver/ExecuteCommandRoutine';
import RunInWorkspacesRoutine from './RunInWorkspacesRoutine';
import filterArgs from '../helpers/filterArgs';
import { EXECUTE_OPTIONS } from '../constants';

export default class ExecuteDriverRoutine extends RunInWorkspacesRoutine<DriverContext> {
  pipeRoutine(context: DriverContext, packageName?: string, packageRoot?: string) {
    if (packageName) {
      this.pipeParallelBuilds(context, packageName, {
        forceConfigOption: true,
        packageRoot,
      });
    } else {
      this.pipeParallelBuilds(context, context.primaryDriver.name);
    }
  }

  /**
   * When a parallel pipe "//" is defined, we need to create an additional routine
   * for each instance.
   */
  pipeParallelBuilds(
    context: DriverContext,
    key: string,
    options: Partial<ExecuteCommandOptions> = {},
  ) {
    const { argv, parallelArgv, primaryDriver } = context;
    const { filteredArgv } = filterArgs(argv, {
      block: EXECUTE_OPTIONS,
    });
    const command = `${primaryDriver.metadata.bin} ${filteredArgv.join(' ')}`.trim();

    this.routines.push(
      new ExecuteCommandRoutine(key, command, {
        ...options,
        argv: filteredArgv,
        tool: this.options.tool,
      }),
    );

    parallelArgv.forEach((pargv) => {
      this.routines.push(
        new ExecuteCommandRoutine(key, `${command} ${pargv.join(' ')}`.trim(), {
          ...options,
          additionalArgv: pargv,
          argv: filteredArgv,
          tool: this.options.tool,
        }),
      );
    });
  }
}
