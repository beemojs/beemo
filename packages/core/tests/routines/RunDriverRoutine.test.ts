/* eslint-disable jest/expect-expect */

import { Path, Project } from '@boost/common';
import { getFixturePath } from '@boost/test-utils';
import { DriverContext } from '../../src/contexts/DriverContext';
import { Driver } from '../../src/Driver';
import { ExecuteCommandOptions } from '../../src/routines/driver/ExecuteCommandRoutine';
import { RunDriverRoutine } from '../../src/routines/RunDriverRoutine';
import { AnyRoutine } from '../../src/routines/RunInWorkspacesRoutine';
import { mockDebugger, mockDriver, mockTool, stubDriverContext } from '../../src/test';

jest.mock('execa');

describe('RunDriverRoutine', () => {
  let routine: RunDriverRoutine;
  let context: DriverContext;
  let driver: Driver;

  function expectPipedRoutines(
    routines: AnyRoutine[],
    tests: (Partial<ExecuteCommandOptions> & { key?: string; title: string })[],
  ) {
    expect(routines).toHaveLength(tests.length);

    tests.forEach((test, i) => {
      const { key, title, ...options } = test;

      if (key) {
        expect(routines[i]).toEqual(
          expect.objectContaining({
            key,
            title,
            options: expect.objectContaining(options),
          }),
        );
      } else {
        expect(routines[i]).toEqual(
          expect.objectContaining({
            title,
            options: expect.objectContaining(options),
          }),
        );
      }
    });
  }

  beforeEach(() => {
    const tool = mockTool();

    driver = mockDriver('primary', tool);
    context = stubDriverContext(driver);

    routine = new RunDriverRoutine('driver', 'Executing driver', { tool });
    // @ts-expect-error Overwrite readonly
    routine.debug = mockDebugger();
  });

  describe('bootstrap()', () => {
    it('adds a routine for the primary driver', async () => {
      await routine.execute(context);

      expectPipedRoutines(routine.routines, [{ title: 'primary -a --foo bar baz' }]);
    });

    it('adds multiple routines when parallel is used', async () => {
      context.parallelArgv = [
        ['--one', '--two=2'],
        ['--three', '-f'],
      ];

      await routine.execute(context);

      expectPipedRoutines(routine.routines, [
        { title: 'primary -a --foo bar baz' },
        { title: 'primary -a --foo bar baz --one --two=2', additionalArgv: ['--one', '--two=2'] },
        { title: 'primary -a --foo bar baz --three -f', additionalArgv: ['--three', '-f'] },
      ]);
    });

    it('adds a routine if parallel is empty', async () => {
      context.parallelArgv = [];

      await routine.execute(context);

      expectPipedRoutines(routine.routines, [{ title: 'primary -a --foo bar baz' }]);
    });

    describe('workspaces', () => {
      const fixturePath = new Path(getFixturePath('workspaces-driver'));

      beforeEach(() => {
        context.args.options.workspaces = '*';
        context.workspaces = ['packages/*'];
        context.workspaceRoot = fixturePath;
        context.cwd = fixturePath;

        // @ts-expect-error Overwrite readonly
        routine.options.tool.project = new Project(fixturePath);
      });

      it('adds a routine for each', async () => {
        await routine.execute(context);

        expectPipedRoutines(routine.routines, [
          {
            key: 'bar',
            title: 'primary -a --foo bar baz',
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/bar').path(),
          },
          {
            key: 'baz',
            title: 'primary -a --foo bar baz',
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/baz').path(),
          },
          {
            key: 'foo',
            title: 'primary -a --foo bar baz',
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/foo').path(),
          },
        ]);
      });

      it('adds a routine for each when parallel is used', async () => {
        context.parallelArgv = [
          ['--one', '--two=2'],
          ['--three', '-f'],
        ];

        await routine.execute(context);

        expectPipedRoutines(routine.routines, [
          {
            key: 'bar',
            title: 'primary -a --foo bar baz',
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/bar').path(),
          },
          {
            key: 'bar',
            title: 'primary -a --foo bar baz --one --two=2',
            additionalArgv: ['--one', '--two=2'],
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/bar').path(),
          },
          {
            key: 'bar',
            title: 'primary -a --foo bar baz --three -f',
            additionalArgv: ['--three', '-f'],
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/bar').path(),
          },
          {
            key: 'baz',
            title: 'primary -a --foo bar baz',
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/baz').path(),
          },
          {
            key: 'baz',
            title: 'primary -a --foo bar baz --one --two=2',
            additionalArgv: ['--one', '--two=2'],
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/baz').path(),
          },
          {
            key: 'baz',
            title: 'primary -a --foo bar baz --three -f',
            additionalArgv: ['--three', '-f'],
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/baz').path(),
          },
          {
            key: 'foo',
            title: 'primary -a --foo bar baz',
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/foo').path(),
          },
          {
            key: 'foo',
            title: 'primary -a --foo bar baz --one --two=2',
            additionalArgv: ['--one', '--two=2'],
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/foo').path(),
          },
          {
            key: 'foo',
            title: 'primary -a --foo bar baz --three -f',
            additionalArgv: ['--three', '-f'],
            forceConfigOption: true,
            packageRoot: fixturePath.append('./packages/foo').path(),
          },
        ]);
      });

      it('errors if workspaces config is empty', async () => {
        await expect(() => {
          context.workspaces = [];

          return routine.execute(context);
        }).rejects.toThrowErrorMatchingSnapshot();
      });
    });
  });
});
