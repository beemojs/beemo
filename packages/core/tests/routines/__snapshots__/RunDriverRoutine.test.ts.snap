// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunDriverRoutine bootstrap() workspaces errors if workspaces config is empty 1`] = `"Option --workspaces=* provided but project is not workspaces enabled."`;
